src/apps/app_bulk_upload.erl

```diff
@@ -0,0 +1,208 @@
+
+upload_blob(Wallet, Blob) when is_binary(Blob) ->
+	upload_blob(whereis(http_entrypoint_node), Wallet, Blob).
+
+upload_blob(Node, Wallet, Blob) ->
+	Queue = app_queue:start(Node, Wallet),
+	Hash = crypto:hash(?BLOB_HASH_ALGO, Blob),
+	upload_chunks(Queue, Hash, Blob),
+	Queue.
+
+%% @doc Takes a binary blob and processes it chunk by chunk. Each chunk is converted into
+%% a transaction and put into the queue. Chunk size is 1MB.
+upload_chunks(Queue, Hash, Blob) ->
+	BlobSize = byte_size(Blob),
+	HasRemainder = BlobSize rem ?CHUNK_SIZE =/= 0,
+	NumberOfChunks = BlobSize div ?CHUNK_SIZE + case HasRemainder of true -> 1; false -> 0 end,
+	upload_chunks(Queue, Hash, Blob, NumberOfChunks, 1).
+
+upload_chunks(Queue, Hash, Blob, NumberOfChunks, ChunkPosition) ->
+	case byte_size(Blob) =< ?CHUNK_SIZE of
+		true ->
+			app_queue:add(Queue, chunk_to_tx(Hash, Blob, NumberOfChunks, ChunkPosition));
+		false ->
+			<< Chunk:?CHUNK_SIZE/binary, Rest/binary >> = Blob,
+			app_queue:add(Queue, chunk_to_tx(Hash, Chunk, NumberOfChunks, ChunkPosition)),
+			upload_chunks(Queue, Hash, Rest, NumberOfChunks, ChunkPosition + 1)
+	end.
+
+%% @doc Converts the given binary chunk into a transaction. A hash of the whole block the chunk
+%% is part of is assigned as a tag.
+chunk_to_tx(Hash, Chunk, NumberOfChunks, ChunkPosition) ->
+	#tx {
+		tags =
+			[
+				{"app_name", "BulkUpload"},
+				{"blob_hash", ar_util:encode(Hash)},
+				{"number_of_chunks", integer_to_binary(NumberOfChunks)},
+				{"chunk_position", integer_to_binary(ChunkPosition)}
+			],
+		data = Chunk
+	}.
+
+%% @doc Searches the local storage for the chunks of the blob with the given hash.
+%% If the blob is reconstructed successfuly, writes it to the specified destination.
+%% The provided hash has to be a Base 64 encoded SHA 256 hash of the file as a binary string.
+download(Hash, Filename) ->
+	{ok, Blob} = download(Hash),
+	file:write_file(Filename, iolist_to_binary(Blob), [write]).
+
+download(Hash) ->
+	TXIDs = app_search:get_entries(<< "blob_hash" >>, Hash),
+	Transactions = lists:map(fun(TX) -> ar_storage:read_tx(TX) end, TXIDs),
+	case lists:member(unavailable, Transactions) of
+		true ->
+			{error, invalid_storage_state};
+		false ->
+			{ok, Blob} = reconstruct_blob(Transactions),
+			BlobHash = ar_util:encode(crypto:hash(?BLOB_HASH_ALGO, iolist_to_binary(Blob))),
+			case BlobHash == Hash of
+				true ->
+					{ok, Blob};
+				false ->
+					{error, invalid_upload}
+			end
+	end.
+
+reconstruct_blob([]) -> {error, not_found};
+reconstruct_blob(Transactions) ->
+	Head = hd(Transactions),
+	{_, NumberOfChunksBinary} = lists:keyfind(<< "number_of_chunks" >>, 1, Head#tx.tags),
+	NumberOfChunks = binary_to_integer(NumberOfChunksBinary),
+	case length(Transactions) < NumberOfChunks of
+		true ->
+			{error, missing_chunks};
+		false ->
+			Sorted = sort_transactions_by_chunk_position(Transactions),
+			SortedUnique = lists:foldr(fun(V, A) -> drop_duplicates(V, A) end, [], Sorted),
```

Comment:
You can also write this like this:
`lists:foldr(fun drop_duplicates/2, [], Sorted)`